diff -ruN minetest-master/src/object_properties.cpp minetest-freeminer/src/object_properties.cpp
--- minetest-master/src/object_properties.cpp	2016-09-03 16:15:52.689779000 -0700
+++ minetest-freeminer/src/object_properties.cpp	2016-09-03 15:44:00.002741000 -0700
@@ -1,20 +1,23 @@
 /*
-Minetest
+object_properties.cpp
 Copyright (C) 2013 celeron55, Perttu Ahola <celeron55@gmail.com>
+*/
+
+/*
+This file is part of Freeminer.
 
-This program is free software; you can redistribute it and/or modify
-it under the terms of the GNU Lesser General Public License as published by
-the Free Software Foundation; either version 2.1 of the License, or
+Freeminer is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
-This program is distributed in the hope that it will be useful,
+Freeminer  is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-GNU Lesser General Public License for more details.
+GNU General Public License for more details.
 
-You should have received a copy of the GNU Lesser General Public License along
-with this program; if not, write to the Free Software Foundation, Inc.,
-51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+You should have received a copy of the GNU General Public License
+along with Freeminer.  If not, see <http://www.gnu.org/licenses/>.
 */
 
 #include "object_properties.h"
@@ -43,12 +46,13 @@
 	stepheight(0),
 	automatic_face_movement_dir(false),
 	automatic_face_movement_dir_offset(0.0),
+	force_load(false),
 	backface_culling(true),
 	nametag(""),
 	nametag_color(255, 255, 255, 255),
 	automatic_face_movement_max_rotation_per_sec(-1)
 {
-	textures.push_back("unknown_object.png");
+	textures.push_back("blank.png");
 	colors.push_back(video::SColor(255,255,255,255));
 }
 
@@ -78,6 +82,7 @@
 	os<<", is_visible="<<is_visible;
 	os<<", makes_footstep_sound="<<makes_footstep_sound;
 	os<<", automatic_rotate="<<automatic_rotate;
+	os<<", force_load="<<force_load;
 	os<<", backface_culling="<<backface_culling;
 	os << ", nametag=" << nametag;
 	os << ", nametag_color=" << "\"" << nametag_color.getAlpha() << "," << nametag_color.getRed()
@@ -115,6 +120,10 @@
 	writeU8(os, automatic_face_movement_dir);
 	writeF1000(os, automatic_face_movement_dir_offset);
 	writeU8(os, backface_culling);
+
+	//freeminer:
+	//writeU8(os, force_load);
+
 	os << serializeString(nametag);
 	writeARGB8(os, nametag_color);
 	writeF1000(os, automatic_face_movement_max_rotation_per_sec);
@@ -157,6 +166,10 @@
 			automatic_face_movement_dir = readU8(is);
 			automatic_face_movement_dir_offset = readF1000(is);
 			backface_culling = readU8(is);
+
+			//freeminer:
+			//force_load = readU8(is);
+
 			nametag = deSerializeString(is);
 			nametag_color = readARGB8(is);
 			automatic_face_movement_max_rotation_per_sec = readF1000(is);
