diff -ruN minetest-master/builtin/mainmenu/dlg_create_world.lua minetest-freeminer/builtin/mainmenu/dlg_create_world.lua
--- minetest-master/builtin/mainmenu/dlg_create_world.lua	2016-09-03 16:15:52.617779000 -0700
+++ minetest-freeminer/builtin/mainmenu/dlg_create_world.lua	2016-09-03 15:43:59.946741000 -0700
@@ -33,6 +33,7 @@
 	end
 	mglist = mglist:sub(1, -2)
 	
+	gamemgr.update_gamelist()
 	local gameid = core.setting_get("menu_last_game")
 	
 	local game, gameidx = nil , 0
@@ -46,7 +47,8 @@
 
 	current_seed = core.formspec_escape(current_seed)
 	local retval =
-		"size[11.5,6.5,true]" ..
+		"size[12,5,false]" ..
+		--"size[11.5,6.5,true]" ..
 		"label[2,0;" .. fgettext("World name") .. "]"..
 		"field[4.5,0.4;6,0.5;te_world_name;;]" ..
 
@@ -57,12 +59,14 @@
 		"dropdown[4.2,2;6.3;dd_mapgen;" .. mglist .. ";" .. selindex .. "]" ..
 
 		"label[2,3;" .. fgettext("Game") .. "]"..
-		"textlist[4.2,3;5.8,2.3;games;" .. gamemgr.gamelist() ..
+		"textlist[4.2,3;5.8,1;games;" .. gamemgr.gamelist() ..
 		";" .. gameidx .. ";true]" ..
 
-		"button[3.25,6;2.5,0.5;world_create_confirm;" .. fgettext("Create") .. "]" ..
-		"button[5.75,6;2.5,0.5;world_create_cancel;" .. fgettext("Cancel") .. "]"
-		
+		"button[7.7,4.5;2.6,0.5;world_create_confirm;" .. fgettext("Create") .. "]" ..
+		"button[4.2,4.5;2.8,0.5;world_create_cancel;" .. fgettext("Cancel") .. "]"
+		--"button[3.25,6;2.5,0.5;world_create_confirm;" .. fgettext("Create") .. "]" ..
+		--"button[5.75,6;2.5,0.5;world_create_cancel;" .. fgettext("Cancel") .. "]"
+
 	if #gamemgr.games == 0 then
 		retval = retval .. "box[2,4;8,1;#ff8800]label[2.25,4;" ..
 				fgettext("You have no subgames installed.") .. "]label[2.25,4.4;" ..
