diff -ruN minetest-master/cmake/Modules/FindIrrlicht.cmake minetest-freeminer/cmake/Modules/FindIrrlicht.cmake
--- minetest-master/cmake/Modules/FindIrrlicht.cmake	2016-09-03 16:15:52.593779000 -0700
+++ minetest-freeminer/cmake/Modules/FindIrrlicht.cmake	2016-09-03 15:43:59.750740000 -0700
@@ -14,7 +14,11 @@
 
 	if(WIN32)
 		if(MSVC)
+			if(CMAKE_CL_64)
+				set(IRRLICHT_SOURCE_DIR_LIBS "${IRRLICHT_SOURCE_DIR}/lib/Win64-visualstudio")
+			else()
 			set(IRRLICHT_SOURCE_DIR_LIBS "${IRRLICHT_SOURCE_DIR}/lib/Win32-visualstudio")
+			endif()
 			set(IRRLICHT_LIBRARY_NAMES Irrlicht.lib)
 		else()
 			set(IRRLICHT_SOURCE_DIR_LIBS "${IRRLICHT_SOURCE_DIR}/lib/Win32-gcc")
