diff -ruN minetest-master/src/game.h minetest-freeminer/src/game.h
--- minetest-master/src/game.h	2016-09-03 16:15:52.497778000 -0700
+++ minetest-freeminer/src/game.h	2016-09-03 15:43:59.870740000 -0700
@@ -1,20 +1,23 @@
 /*
-Minetest
+game.h
 Copyright (C) 2013 celeron55, Perttu Ahola <celeron55@gmail.com>
+*/
+
+/*
+This file is part of Freeminer.
 
-This program is free software; you can redistribute it and/or modify
-it under the terms of the GNU Lesser General Public License as published by
-the Free Software Foundation; either version 2.1 of the License, or
+Freeminer is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
-This program is distributed in the hope that it will be useful,
+Freeminer  is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-GNU Lesser General Public License for more details.
+GNU General Public License for more details.
 
-You should have received a copy of the GNU Lesser General Public License along
-with this program; if not, write to the Free Software Foundation, Inc.,
-51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+You should have received a copy of the GNU General Public License
+along with Freeminer.  If not, see <http://www.gnu.org/licenses/>.
 */
 
 #ifndef GAME_HEADER
@@ -143,7 +146,7 @@
 class ChatBackend;  /* to avoid having to include chat.h */
 struct SubgameSpec;
 
-void the_game(bool *kill,
+bool the_game(bool *kill,
 		bool random_input,
 		InputHandler *input,
 		IrrlichtDevice *device,
@@ -156,7 +159,9 @@
 		ChatBackend &chat_backend,
 		bool *reconnect_requested,
 		const SubgameSpec &gamespec, // Used for local game
-		bool simple_singleplayer_mode);
+		bool simple_singleplayer_mode,
+		unsigned int autoexit
+	);
 
 #endif
 
