diff -ruN minetest-master/src/clientmedia.cpp minetest-freeminer/src/clientmedia.cpp
--- minetest-master/src/clientmedia.cpp	2016-09-03 16:15:52.597779000 -0700
+++ minetest-freeminer/src/clientmedia.cpp	2016-09-03 15:43:59.754740000 -0700
@@ -1,20 +1,23 @@
 /*
-Minetest
+clientmedia.cpp
 Copyright (C) 2013 celeron55, Perttu Ahola <celeron55@gmail.com>
+*/
+
+/*
+This file is part of Freeminer.
 
-This program is free software; you can redistribute it and/or modify
-it under the terms of the GNU Lesser General Public License as published by
-the Free Software Foundation; either version 2.1 of the License, or
+Freeminer is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
-This program is distributed in the hope that it will be useful,
+Freeminer  is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-GNU Lesser General Public License for more details.
+GNU General Public License for more details.
 
-You should have received a copy of the GNU Lesser General Public License along
-with this program; if not, write to the Free Software Foundation, Inc.,
-51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+You should have received a copy of the GNU General Public License
+along with Freeminer.  If not, see <http://www.gnu.org/licenses/>.
 */
 
 #include "clientmedia.h"
@@ -182,6 +185,12 @@
 		FileStatus *filestatus = it->second;
 		const std::string &sha1 = filestatus->sha1;
 
+		if (g_settings->get("video_driver") == "null") {
+			filestatus->received = true;
+			m_uncached_count--;
+			continue;
+		}
+
 		std::ostringstream tmp_os(std::ios_base::binary);
 		bool found_in_cache = m_media_cache.load(hex_encode(sha1), tmp_os);
 
