diff -ruN minetest-master/src/util/directiontables.h minetest-freeminer/src/util/directiontables.h
--- minetest-master/src/util/directiontables.h	2016-09-03 16:15:52.705779000 -0700
+++ minetest-freeminer/src/util/directiontables.h	2016-09-03 15:43:59.826740000 -0700
@@ -1,20 +1,23 @@
 /*
-Minetest
+util/directiontables.h
 Copyright (C) 2010-2013 celeron55, Perttu Ahola <celeron55@gmail.com>
+*/
+
+/*
+This file is part of Freeminer.
 
-This program is free software; you can redistribute it and/or modify
-it under the terms of the GNU Lesser General Public License as published by
-the Free Software Foundation; either version 2.1 of the License, or
+Freeminer is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
-This program is distributed in the hope that it will be useful,
+Freeminer  is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-GNU Lesser General Public License for more details.
+GNU General Public License for more details.
 
-You should have received a copy of the GNU Lesser General Public License along
-with this program; if not, write to the Free Software Foundation, Inc.,
-51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+You should have received a copy of the GNU General Public License
+along with Freeminer.  If not, see <http://www.gnu.org/licenses/>.
 */
 
 #ifndef UTIL_DIRECTIONTABLES_HEADER
@@ -23,7 +26,7 @@
 #include "../irrlichttypes.h"
 #include "../irr_v3d.h"
 
-extern const v3s16 g_6dirs[6];
+extern const v3s16 g_6dirs[7];
 
 extern const v3s16 g_26dirs[26];
 
