diff -ruN minetest-master/builtin/init.lua minetest-freeminer/builtin/init.lua
--- minetest-master/builtin/init.lua	2016-09-03 16:15:52.589779000 -0700
+++ minetest-freeminer/builtin/init.lua	2016-09-03 15:43:59.750740000 -0700
@@ -23,6 +23,8 @@
 math.randomseed(os.time())
 os.setlocale("C", "numeric")
 minetest = core
+freeminer = core
+multicraft = core
 
 -- Load other files
 local scriptdir = core.get_builtin_path() .. DIR_DELIM
@@ -30,10 +32,19 @@
 local commonpath = scriptdir .. "common" .. DIR_DELIM
 local asyncpath = scriptdir .. "async" .. DIR_DELIM
 
+--dofile(scriptdir .. "profiler.lua") --TODO: repair me
+
+--[[ too buggy
 dofile(commonpath .. "strict.lua")
+]]
+
 dofile(commonpath .. "serialize.lua")
 dofile(commonpath .. "misc_helpers.lua")
 
+dofile(scriptdir.."key_value_storage.lua")
+
+--PLATFORM = "Android" -- for test
+
 if INIT == "game" then
 	dofile(gamepath .. "init.lua")
 elseif INIT == "mainmenu" then
@@ -41,7 +52,11 @@
 	if mainmenuscript ~= nil and mainmenuscript ~= "" then
 		dofile(mainmenuscript)
 	else
+	  if PLATFORM == "Android" then
 		dofile(core.get_mainmenu_path() .. DIR_DELIM .. "init.lua")
+	  else
+		dofile(core.get_mainmenu_path() .. DIR_DELIM .. "fm_init.lua")
+	  end
 	end
 elseif INIT == "async" then
 	dofile(asyncpath .. "init.lua")
