diff -ruN "minetest-master/build/android_fm/worlds/World 7/map_meta.json" "minetest-freeminer/build/android_fm/worlds/World 7/map_meta.json"
--- "minetest-master/build/android_fm/worlds/World 7/map_meta.json"	1969-12-31 16:00:00.000000000 -0800
+++ "minetest-freeminer/build/android_fm/worlds/World 7/map_meta.json"	2016-09-03 15:43:59.526738000 -0700
@@ -0,0 +1,3 @@
+{
+	"mg_name" : "v7"
+}
\ No newline at end of file
