diff -ruN minetest-master/src/mapgen_fractal.cpp minetest-freeminer/src/mapgen_fractal.cpp
--- minetest-master/src/mapgen_fractal.cpp	2016-09-03 16:15:52.501778000 -0700
+++ minetest-freeminer/src/mapgen_fractal.cpp	2016-09-03 15:43:59.874740000 -0700
@@ -102,7 +102,7 @@
 }
 
 
-void MapgenFractalParams::readParams(const Settings *settings)
+void MapgenFractalParams::readParams(Settings *settings)
 {
 	settings->getFlagStrNoEx("mgfractal_spflags",  spflags, flagdesc_mapgen_fractal);
 	settings->getFloatNoEx("mgfractal_cave_width", cave_width);
@@ -226,7 +226,7 @@
 
 	//printf("makeChunk: %dms\n", t.stop());
 
-	updateLiquid(&data->transforming_liquid, full_node_min, full_node_max);
+	updateLiquid(full_node_min, full_node_max);
 
 	if (flags & MG_LIGHT)
 		calcLighting(node_min - v3s16(0, 1, 0), node_max + v3s16(0, 1, 0),
