diff -ruN minetest-master/src/shader.cpp minetest-freeminer/src/shader.cpp
--- minetest-master/src/shader.cpp	2016-09-03 16:15:52.665779000 -0700
+++ minetest-freeminer/src/shader.cpp	2016-09-03 15:43:59.802740000 -0700
@@ -1,21 +1,24 @@
 /*
-Minetest
+shader.cpp
 Copyright (C) 2013 celeron55, Perttu Ahola <celeron55@gmail.com>
 Copyright (C) 2013 Kahrl <kahrl@gmx.net>
+*/
+
+/*
+This file is part of Freeminer.
 
-This program is free software; you can redistribute it and/or modify
-it under the terms of the GNU Lesser General Public License as published by
-the Free Software Foundation; either version 2.1 of the License, or
+Freeminer is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
-This program is distributed in the hope that it will be useful,
+Freeminer  is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-GNU Lesser General Public License for more details.
+GNU General Public License for more details.
 
-You should have received a copy of the GNU Lesser General Public License along
-with this program; if not, write to the Free Software Foundation, Inc.,
-51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+You should have received a copy of the GNU General Public License
+along with Freeminer.  If not, see <http://www.gnu.org/licenses/>.
 */
 
 #include <fstream>
@@ -23,6 +26,7 @@
 #include "shader.h"
 #include "irrlichttypes_extrabloated.h"
 #include "debug.h"
+//#include <fstream>
 #include "filesys.h"
 #include "util/container.h"
 #include "util/thread.h"
