diff -ruN minetest-master/builtin/mainmenu/init.lua minetest-freeminer/builtin/mainmenu/init.lua
--- minetest-master/builtin/mainmenu/init.lua	2016-09-03 16:15:52.613779000 -0700
+++ minetest-freeminer/builtin/mainmenu/init.lua	2016-09-03 15:43:59.942741000 -0700
@@ -41,19 +41,23 @@
 
 dofile(menupath .. DIR_DELIM .. "dlg_config_world.lua")
 dofile(menupath .. DIR_DELIM .. "dlg_settings_advanced.lua")
-if PLATFORM ~= "Android" then
+--if PLATFORM ~= "Android" then
 	dofile(menupath .. DIR_DELIM .. "dlg_create_world.lua")
 	dofile(menupath .. DIR_DELIM .. "dlg_delete_mod.lua")
 	dofile(menupath .. DIR_DELIM .. "dlg_delete_world.lua")
 	dofile(menupath .. DIR_DELIM .. "dlg_rename_modpack.lua")
+--[[
 end
-
+]]
 local tabs = {}
 
 tabs.settings = dofile(menupath .. DIR_DELIM .. "tab_settings.lua")
 tabs.mods = dofile(menupath .. DIR_DELIM .. "tab_mods.lua")
 tabs.credits = dofile(menupath .. DIR_DELIM .. "tab_credits.lua")
+--[[
 if PLATFORM == "Android" then
+]]
+if false then
 	tabs.simple_main = dofile(menupath .. DIR_DELIM .. "tab_simple_main.lua")
 else
 	tabs.singleplayer = dofile(menupath .. DIR_DELIM .. "tab_singleplayer.lua")
@@ -61,7 +65,6 @@
 	tabs.server = dofile(menupath .. DIR_DELIM .. "tab_server.lua")
 	tabs.texturepacks = dofile(menupath .. DIR_DELIM .. "tab_texturepacks.lua")
 end
-
 --------------------------------------------------------------------------------
 local function main_event_handler(tabview, event)
 	if event == "MenuQuit" then
@@ -75,7 +78,10 @@
 	-- Init gamedata
 	gamedata.worldindex = 0
 
+--[[
 	if PLATFORM == "Android" then
+]]
+	if false then
 		local world_list = core.get_worlds()
 		local world_index
 
@@ -120,7 +126,7 @@
 		menudata.worldlist:set_sortmode("alphabetic")
 
 		if not core.setting_get("menu_last_game") then
-			local default_game = core.setting_get("default_game") or "minetest"
+			local default_game = core.setting_get("default_game") or "default"
 			core.setting_set("menu_last_game", default_game)
 		end
 
@@ -130,16 +136,21 @@
 	-- Create main tabview
 	local tv_main = tabview_create("maintab", {x = 12, y = 5.2}, {x = 0, y = 0})
 
+--[[
 	if PLATFORM == "Android" then
 		tv_main:add(tabs.simple_main)
 		tv_main:add(tabs.settings)
 	else
+]]
+	if true then
 		tv_main:set_autosave_tab(true)
 		tv_main:add(tabs.singleplayer)
 		tv_main:add(tabs.multiplayer)
 		tv_main:add(tabs.server)
 		tv_main:add(tabs.settings)
+		if PLATFORM ~= "Android" then
 		tv_main:add(tabs.texturepacks)
+		end
 	end
 
 	tv_main:add(tabs.mods)
@@ -148,8 +159,11 @@
 	tv_main:set_global_event_handler(main_event_handler)
 	tv_main:set_fixed_size(false)
 
+--[[
 	if PLATFORM ~= "Android" then
-		tv_main:set_tab(core.setting_get("maintab_LAST"))
+]]
+	if true then
+		tv_main:set_tab(core.setting_get("maintab_LAST") or 'multiplayer')
 	end
 	ui.set_default("maintab")
 	tv_main:show()
@@ -163,6 +177,8 @@
 
 	ui.update()
 
+	updater_init()
+
 	core.sound_play("main_menu", true)
 end
 
