diff -ruN minetest-master/src/threading/thread_local.h minetest-freeminer/src/threading/thread_local.h
--- minetest-master/src/threading/thread_local.h	1969-12-31 16:00:00.000000000 -0800
+++ minetest-freeminer/src/threading/thread_local.h	2016-09-03 15:43:59.882740000 -0700
@@ -0,0 +1,30 @@
+/*
+This file is part of Freeminer.
+
+Freeminer is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+Freeminer  is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with Freeminer.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef THREADING_THREAD_LOCAL
+#define THREADING_THREAD_LOCAL
+
+#include "../config.h"
+
+#if HAVE_THREAD_LOCAL
+#define THREAD_LOCAL thread_local
+#else
+#define THREAD_LOCAL
+#endif
+
+
+#endif
