diff -ruN minetest-master/src/mapgen_v6.h minetest-freeminer/src/mapgen_v6.h
--- minetest-master/src/mapgen_v6.h	2016-09-03 16:15:52.637779000 -0700
+++ minetest-freeminer/src/mapgen_v6.h	2016-09-03 15:43:59.782740000 -0700
@@ -1,20 +1,23 @@
 /*
-Minetest
+mapgen_v6.h
 Copyright (C) 2010-2015 celeron55, Perttu Ahola <celeron55@gmail.com>
+*/
+
+/*
+This file is part of Freeminer.
 
-This program is free software; you can redistribute it and/or modify
-it under the terms of the GNU Lesser General Public License as published by
-the Free Software Foundation; either version 2.1 of the License, or
+Freeminer is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
-This program is distributed in the hope that it will be useful,
+Freeminer  is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-GNU Lesser General Public License for more details.
+GNU General Public License for more details.
 
-You should have received a copy of the GNU Lesser General Public License along
-with this program; if not, write to the Free Software Foundation, Inc.,
-51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+You should have received a copy of the GNU General Public License
+along with Freeminer.  If not, see <http://www.gnu.org/licenses/>.
 */
 
 #ifndef MAPGENV6_HEADER
@@ -72,7 +75,7 @@
 	MapgenV6Params();
 	~MapgenV6Params() {}
 
-	void readParams(const Settings *settings);
+	void readParams(Settings *settings);
 	void writeParams(Settings *settings) const;
 };
 
@@ -118,7 +121,8 @@
 	content_t c_dirt_with_snow;
 	content_t c_snow;
 	content_t c_snowblock;
-	content_t c_ice;
+
+	content_t c_ice, c_clay;
 
 	content_t c_cobble;
 	content_t c_mossycobble;
@@ -143,25 +147,26 @@
 	bool block_is_underground(u64 seed, v3s16 blockpos);
 	s16 find_ground_level_from_noise(u64 seed, v2s16 p2d, s16 precision);
 
-	float getHumidity(v2s16 p);
+	float getHumidity(v3POS p);
 	float getTreeAmount(v2s16 p);
 	bool getHaveAppleTree(v2s16 p);
 	float getMudAmount(v2s16 p);
 	virtual float getMudAmount(int index);
 	bool getHaveBeach(v2s16 p);
 	bool getHaveBeach(int index);
-	BiomeV6Type getBiome(v2s16 p);
-	BiomeV6Type getBiome(int index, v2s16 p);
+	BiomeV6Type getBiome(v3POS p);
+	BiomeV6Type getBiome(int index, v3POS p);
 
 	u32 get_blockseed(u64 seed, v3s16 p);
 
 	virtual void calculateNoise();
-	int generateGround();
+	virtual int generateGround();
 	void addMud();
 	void flowMud(s16 &mudflow_minpos, s16 &mudflow_maxpos);
 	void growGrass();
 	void placeTreesAndJungleGrass();
 	virtual void generateCaves(int max_stone_y);
+	virtual void generateExperimental() {}
 };
 
 #endif
