diff -ruN minetest-master/src/client/inputhandler.h minetest-freeminer/src/client/inputhandler.h
--- minetest-master/src/client/inputhandler.h	2016-09-03 16:15:52.497778000 -0700
+++ minetest-freeminer/src/client/inputhandler.h	2016-09-03 15:43:59.510738000 -0700
@@ -29,6 +29,7 @@
 	// This is the one method that we have to implement
 	virtual bool OnEvent(const SEvent& event)
 	{
+		try {
 		/*
 			React to nothing here if a menu is active
 		*/
@@ -111,6 +112,9 @@
 				std::string("Irrlicht: ") + (const char*) event.LogEvent.Text);
 			return true;
 		}
+		} catch(std::exception &e) {
+			errorstream<<"input exception: " << e.what() << std::endl;
+		}
 		/* always return false in order to continue processing events */
 		return false;
 	}
