diff -ruN minetest-master/src/voxel.h minetest-freeminer/src/voxel.h
--- minetest-master/src/voxel.h	2016-09-03 16:15:52.681779000 -0700
+++ minetest-freeminer/src/voxel.h	2016-09-03 15:43:59.994741000 -0700
@@ -1,20 +1,23 @@
 /*
-Minetest
+voxel.h
 Copyright (C) 2013 celeron55, Perttu Ahola <celeron55@gmail.com>
+*/
+
+/*
+This file is part of Freeminer.
 
-This program is free software; you can redistribute it and/or modify
-it under the terms of the GNU Lesser General Public License as published by
-the Free Software Foundation; either version 2.1 of the License, or
+Freeminer is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
-This program is distributed in the hope that it will be useful,
+Freeminer  is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-GNU Lesser General Public License for more details.
+GNU General Public License for more details.
 
-You should have received a copy of the GNU Lesser General Public License along
-with this program; if not, write to the Free Software Foundation, Inc.,
-51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+You should have received a copy of the GNU General Public License
+along with Freeminer.  If not, see <http://www.gnu.org/licenses/>.
 */
 
 #ifndef VOXEL_HEADER
@@ -47,12 +50,6 @@
 */
 
 /*
-	Debug stuff
-*/
-extern u32 emerge_time;
-extern u32 emerge_load_time;
-
-/*
 	This class resembles aabbox3d<s16> a lot, but has inclusive
 	edges for saner handling of integer sizes
 */
@@ -135,7 +132,8 @@
 
 	v3s16 getExtent() const
 	{
-		return MaxEdge - MinEdge + v3s16(1,1,1);
+		static const auto v3POS1 = v3s16(1,1,1);
+		return MaxEdge - MinEdge + v3POS1;
 	}
 
 	/* Because MaxEdge and MinEdge are included in the voxel area an empty extent
@@ -592,6 +590,15 @@
 	*/
 	//bool m_disable_water_climb;
 
+	//freeminer:
+	// for Map compatibility:
+	const MapNode & getNodeTry(const v3s16 &p) {
+		if (m_area.contains(p))
+			return getNodeRefUnsafe(p);
+		return ContentIgnoreNode;
+	}
+
+
 private:
 };
 
