diff -ruN minetest-master/build/debian/freeminer-data.install minetest-freeminer/build/debian/freeminer-data.install
--- minetest-master/build/debian/freeminer-data.install	1969-12-31 16:00:00.000000000 -0800
+++ minetest-freeminer/build/debian/freeminer-data.install	2016-09-03 15:43:59.766740000 -0700
@@ -0,0 +1,11 @@
+usr/share/doc/freeminer/mapformat.txt usr/share/doc/freeminer-data
+usr/share/doc/freeminer/lua_api.txt   usr/share/doc/freeminer-data
+usr/share/doc/freeminer/README.md     usr/share/doc/freeminer-data
+usr/share/doc/freeminer/freeminer.conf.example usr/share/doc/freeminer-data
+
+usr/share/games/freeminer/builtin/*   usr/share/games/freeminer/builtin
+usr/share/games/freeminer/textures/*  usr/share/games/freeminer/textures
+usr/share/games/freeminer/games/*     usr/share/games/freeminer/games
+usr/share/games/freeminer/client/*    usr/share/games/freeminer/client
+usr/share/locale/*                   usr/share/locale
+
