diff -ruN minetest-master/src/inventory.cpp minetest-freeminer/src/inventory.cpp
--- minetest-master/src/inventory.cpp	2016-09-03 16:15:52.641779000 -0700
+++ minetest-freeminer/src/inventory.cpp	2016-09-03 15:43:59.786740000 -0700
@@ -1,20 +1,23 @@
 /*
-Minetest
+inventory.cpp
 Copyright (C) 2010-2013 celeron55, Perttu Ahola <celeron55@gmail.com>
+*/
+
+/*
+This file is part of Freeminer.
 
-This program is free software; you can redistribute it and/or modify
-it under the terms of the GNU Lesser General Public License as published by
-the Free Software Foundation; either version 2.1 of the License, or
+Freeminer is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
-This program is distributed in the hope that it will be useful,
+Freeminer  is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-GNU Lesser General Public License for more details.
+GNU General Public License for more details.
 
-You should have received a copy of the GNU Lesser General Public License along
-with this program; if not, write to the Free Software Foundation, Inc.,
-51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+You should have received a copy of the GNU General Public License
+along with Freeminer.  If not, see <http://www.gnu.org/licenses/>.
 */
 
 #include "inventory.h"
@@ -622,13 +625,19 @@
 
 const ItemStack& InventoryList::getItem(u32 i) const
 {
-	assert(i < m_size); // Pre-condition
+	if (i >= m_size) {
+		errorstream<<"InventoryList::getItem const: Wrong item requested i="<<i<<" size="<<m_size<<std::endl;
+		return m_items[0]; // BAD
+	}
 	return m_items[i];
 }
 
 ItemStack& InventoryList::getItem(u32 i)
 {
-	assert(i < m_size); // Pre-condition
+	if (i >= m_size) {
+		errorstream<<"InventoryList::getItem: Wrong item requested i="<<i<<" size="<<m_size<<std::endl;
+		return m_items[0]; // BAD
+	}
 	return m_items[i];
 }
 
@@ -645,7 +654,10 @@
 
 void InventoryList::deleteItem(u32 i)
 {
-	assert(i < m_items.size()); // Pre-condition
+	if(i >= m_items.size()) {
+		errorstream<<"InventoryList::deleteItem: Wrong item deleting i="<<i<<" size="<<m_size<<std::endl;
+		return;
+	}
 	m_items[i].clear();
 }
 
