diff -ruN minetest-master/minetest.conf.example minetest-freeminer/minetest.conf.example
--- minetest-master/minetest.conf.example	2016-09-03 16:15:52.661779000 -0700
+++ minetest-freeminer/minetest.conf.example	2016-09-03 15:43:59.970741000 -0700
@@ -1,13 +1,13 @@
-#    This file contains a list of all available settings and their default value for minetest.conf
+#    This file contains a list of all available settings and their default value for freeminer.conf
 
 #    By default, all the settings are commented and not functional.
 #    Uncomment settings by removing the preceding #.
 
-#    minetest.conf is read by default from:
-#    ../minetest.conf
-#    ../../minetest.conf
+#    freeminer.conf is read by default from:
+#    ../freeminer.conf
+#    ../../freeminer.conf
 #    Any other path can be chosen by passing the path as a parameter
-#    to the program, eg. "minetest.exe --config ../minetest.conf.example".
+#    to the program, eg. "freeminer.exe --config ../freeminer.conf.example".
 
 #    Further documentation:
 #    http://wiki.minetest.net/
@@ -284,7 +284,7 @@
 
 #    URL to the server list displayed in the Multiplayer Tab.
 #    type: string
-# serverlist_url = servers.minetest.net
+# serverlist_url = servers.freeminer.org
 
 #    File in client/serverlist/ that contains your favorite servers displayed in the Multiplayer Tab.
 #    type: string
@@ -733,7 +733,7 @@
 
 #    Name of the server, to be displayed when players join and in the serverlist.
 #    type: string
-# server_name = Minetest server
+# server_name = freeminer server
 
 #    Description of server, to be displayed when players join and in the serverlist.
 #    type: string
@@ -741,20 +741,20 @@
 
 #    Domain name of server, to be displayed in the serverlist.
 #    type: string
-# server_address = game.minetest.net
+# server_address = game.freeminer.org
 
 #    Homepage of server, to be displayed in the serverlist.
 #    type: string
-# server_url = http://minetest.net
+# server_url = http://freeminer.org
 
 #    Automaticaly report to the serverlist.
 #    type: bool
 # server_announce = false
 
 #    Announce to this serverlist.
-#    If you want to announce your ipv6 address, use  serverlist_url = v6.servers.minetest.net.
+#    If you want to announce your ipv6 address, use  serverlist_url = v6.servers.freeminer.org
 #    type: string
-# serverlist_url = servers.minetest.net
+# serverlist_url = servers.freeminer.org
 
 #    Disable escape sequences, e.g. chat coloring.
 #    Use this if you want to run a server with pre-0.4.14 clients and you want to disable
@@ -818,7 +818,7 @@
 #    Default game when creating a new world.
 #    This will be overridden when creating a world from the main menu.
 #    type: string
-# default_game = minetest
+# default_game = default
 
 #    Message of the day displayed to players connecting.
 #    type: string
@@ -1041,8 +1041,8 @@
 
 #    Name of map generator to be used when creating a new world.
 #    Creating a world in the main menu will override this.
-#    type: enum values: v5, v6, v7, flat, valleys, fractal, singlenode
-# mg_name = v6
+#    type: enum values: indev, v7, valleys, v5, math, fractal, flat, singlenode, v6
+# mg_name = indev
 
 #    Water surface level of the world.
 #    type: int
@@ -1607,3 +1607,250 @@
 #    type: int
 # profiler_print_interval = 0
 
+
+#
+# Freeminer
+#
+
+## Client
+
+#    Minimum wanted FPS.
+#    The amount of rendered stuff is dynamically set according to this. and viewing range min and max.
+#    type: int
+# wanted_fps = 30
+
+#    The allowed adjustment range for the automatic rendering range adjustment.
+#    Set this to be equal to viewing range minimum to disable the auto-adjustment algorithm.
+#    type: int
+# viewing_range_max = 10000
+
+#    Start optimizing far blocks, in blocks (16 nodes). good values 2 - 5 . 0 = disable
+#    type: int
+# farmesh = 0
+
+#    type: int
+# farmesh_step = 2
+
+#    Set to true to disable wield light (enabled by default, requires shaders)
+#    type: bool
+# disable_wieldlight = false
+
+#    Increases performance on modern GPUs a lot
+#    it may be needed to decrease client_unload_unused_data_timeout to limit the RAM usage
+#    type: bool
+# enable_vbo = false
+
+#    Reduce FOV when speed increases
+#    type: bool
+# movement_fov = true
+
+#    FOV when using zoom, default 15
+#    type: int
+# zoom_fov = 15
+
+#    [dev] Shadows from objects
+#    type: bool
+# shadows = false
+
+#    Increase performance of headless client ( video_driver = null )
+#    type: bool
+# headless_optimize = false
+
+#    Disable death screen, respawn automaticaly
+#    type: bool
+# respawn_auto = false
+
+#    Save password for every server in favorites (in plaintext)
+#    type: bool
+# password_save = true
+
+#    Show cmd console on Windows
+#    type: bool
+# console_enabled = false
+
+#    Path for screenshots
+#    type: string
+# screenshot_path = screenshots
+
+#    In-game chat buffer size (number of messages to show)
+#    type: int
+# chat_buffer_size = 6
+
+#    Selection box border color (R,G,B)
+#    type: string
+# selectionbox_color = (0,0,0)
+
+#    Android only: use better keyboard. glitchy on <4.4 versions
+#    type: enum values: 0, 1, 2
+# android_keyboard = 0
+
+## Server
+
+#    Lan udp announce
+#    type: bool
+# serverlist_lan = true
+
+#    Save auth data to leveldb kv storage
+#    type: bool
+# auth_kv = false
+
+#    If true, blocks are cached (and generated if not before) before a player is spawned.
+#    type: bool
+# cache_block_before_spawn = true
+
+#    Defines the maximum height a player can spawn in a map, above water level
+#    type: int
+# max_spawn_height = 50
+
+#    Allow players with any symbols in names
+#    WARNING!!! SECURITY RISK WITH SOME MODULES
+#    type: bool
+# enable_any_name = false
+
+#    Report broken recipes in mods
+#    type: bool
+# mod_debugging = false
+
+#    if entities are allowed to force load the world
+#    type: bool
+# enable_force_load = true
+
+#    Start profiler when the server starts. Profiler is only available when LuaJIT is used.
+#    Use /profiler_start to start the profiler manually and /profiler_stop to stop it.
+#    The log name profile.txt will only be generated when you manually execute /profiler_stop
+#    type: bool
+# profiler_autostart = false
+
+#    Save freshly generated but not changed block, disable it for reducing map size on big servers
+#    type: bool
+# save_generated_block = true
+
+#    Cpu-eating debug timing tool (also automatically enabled with debug_log_level>=3)
+#    type: int
+# time_taker_enabled = 0
+
+#    Enable thread for send_blocks and thread for map stuff (liquid, map save, ...)  Disable if you have frequent crashes
+#    type: bool
+# more_threads = true
+
+#    Process abms for blocks out of active area, one block per step. Can take 100-200ms per block
+#    type: bool
+# abm_random = false
+
+#    Default privs in creative mode
+#    type: string
+# default_privs_creative = interact, shout, fly, fast
+
+#    Change falling direction in cost of speed. 10 - you are bird, 80 - good wingsuite, 200 - hammer
+#    type: int
+# movement_fall_aerodynamics = 110
+
+#    Constant volume liquids
+#    type: bool
+# liquid_real = false
+
+#    Update liquids every .. recommend for finite: 0.2
+#    type: float
+# liquid_update = 0.4
+
+#    Send to client updated liquid every (increase on high loaded servers with liquid_real)
+#    type: float
+# liquid_send = 1.0
+
+#    Relax flowing blocks to source if level near max and N nearby
+#    source blocks, more realistic, but not true constant.
+#    values: 0,1,2,3,4 : 0 - disable, 1 - most aggresive
+#    (for finite liquids)
+#    type: int
+# liquid_relax = 2
+
+#    Optimization: faster cave flood (and not true constant)
+#    (for finite liquids)
+#    type: bool
+# liquid_fast_flood = true
+
+#    Enable weather (cold-hot, water freeze-melt). use only with liquid_real=1
+#    type: bool
+# weather = true
+
+#    Weather noise maps
+#    type: string
+# mg_biome_np_heat = 15,   30,  (500, 500, 500), 5349, 2, 0.60)
+
+#    type: string
+# mg_biome_np_humidity = 50,   50,  (500, 500, 500), 842,  3, 0.45)
+
+#    Season amplitude in celsius
+#    type: float
+# weather_heat_season = 30
+
+#    Daily amplitude in celsius
+#    type: float
+# weather_heat_daily = 8
+
+#    Period width on map (in nodes)
+#    type: float
+# weather_heat_width = 3000
+
+#    Enable hot core (start at, nodes from bottom), 0 for disable
+#    type: float
+# weather_hot_core = 1000
+
+#    Add 1c on every height (in nodes)  333 = 3c per 1000 nodes
+#    type: float
+# weather_heat_height = -333
+
+#    Bind biomes to weather map in mapgen v6 and indev
+#    type: bool
+# weather_biome = false
+
+#    Length of year in days for seasons change. With default time_speed 365 days = 5 real days for year. 30 days = 10 real hours
+#    type: int
+# year_days = 30
+
+#    Same for humidity:
+#    type: float
+# weather_humidity_season = 30
+
+#    type: float
+# weather_humidity_daily = -12
+
+#    type: float
+# weather_humidity_width = 300
+
+#    Humidity change amplitude
+#    type: float
+# weather_humidity_days = 2
+
+#    Offset, scale, spread factor, seed offset, number of octaves, persistence, farscale, farspread
+#    type: string
+# mgindev_np_terrain_base = -4,   20,  (250, 250, 250), 82341, 5, 0.6,  10,  10
+
+#    type: string
+# mgindev_np_terrain_higher = 20,   16,  (500, 500, 500), 85039, 5, 0.6,  10,  10
+
+#    type: string
+# mgindev_np_steepness = 0.85, 0.5, (125, 125, 125), -932,  5, 0.7,  2,   10
+
+#    type: string
+# mgindev_np_mud = 4,    2,   (200, 200, 200), 91013, 3, 0.55, 1,   1
+
+#    type: string
+# mgindev_np_float_islands1 = 0,    1,   (64,  64,  64 ), 3683,  5, 0.5,  1,   1.5
+
+#    type: string
+# mgindev_np_float_islands2 = 0,    1,   (8,   8,   8  ), 9292,  2, 0.5,  1,   1.5
+
+#    type: string
+# mgindev_np_float_islands3 = 0,    1,   (256, 256, 256), 6412,  2, 0.5,  1,   0.5
+
+#    type: string
+# mgindev_np_biome = 0,    1,   (250, 250, 250), 9130,  3, 0.50, 1,   10
+
+#    Float islands starts from height, 0 to disable
+#    type: int
+# mgindev_float_islands = 500
+
+#    Params for math mapgen in json format. generators supported: sphere, mandelbox, mengersponge, ...      dont forget to lower water_level = -30000
+#    type: string
+# mg_math = {"generator":"sphere"}
Binary files minetest-master/misc/freeminer-24x24.png and minetest-freeminer/misc/freeminer-24x24.png differ
