diff -ruN minetest-master/misc/freeminer.desktop minetest-freeminer/misc/freeminer.desktop
--- minetest-master/misc/freeminer.desktop	1969-12-31 16:00:00.000000000 -0800
+++ minetest-freeminer/misc/freeminer.desktop	2016-09-03 15:43:59.750740000 -0700
@@ -0,0 +1,17 @@
+[Desktop Entry]
+Name=Freeminer
+GenericName=Freeminer
+Comment=Multiplayer infinite-world block sandbox
+Comment[de]=Mehrspieler-Sandkastenspiel mit unendlichen Blockwelten
+Comment[es]=Juego sandbox multijugador con mundos infinitos
+Comment[fr]=Jeu multijoueurs de type bac à sable avec des mondes infinis
+Comment[ja]=マルチプレイに対応した、無限の世界のブロック型サンドボックスゲームです
+Comment[ru]=Игра-песочница с безграничным миром, состоящим из блоков
+Comment[tr]=Tek-Çok oyuncuyla küplerden sonsuz dünyalar inşa et
+Exec=freeminer
+Icon=freeminer
+Terminal=false
+Type=Application
+Categories=Game;Simulation;
+StartupNotify=false
+Keywords=sandbox;world;mining;crafting;blocks;nodes;multiplayer;roleplaying;
Binary files minetest-master/misc/freeminer.ico and minetest-freeminer/misc/freeminer.ico differ
Binary files minetest-master/misc/freeminer-icon.icns and minetest-freeminer/misc/freeminer-icon.icns differ
