diff -ruN minetest-master/builtin/game/chatcommands.lua minetest-freeminer/builtin/game/chatcommands.lua
--- minetest-master/builtin/game/chatcommands.lua	2016-09-03 16:15:52.529778000 -0700
+++ minetest-freeminer/builtin/game/chatcommands.lua	2016-09-03 15:43:59.646739000 -0700
@@ -97,7 +97,7 @@
 		local function format_help_line(cmd, def)
 			local msg = core.colorize("#00ffff", "/"..cmd)
 			if def.params and def.params ~= "" then
-				msg = msg .. " " .. def.params
+				msg = msg .. " " .. core.colorize("#eeeeee", def.params)
 			end
 			if def.description and def.description ~= "" then
 				msg = msg .. ": " .. def.description
@@ -588,6 +588,7 @@
 		if not itemstring then
 			return false, "ItemString required"
 		end
+		core.stat_add("giveme", name)
 		return handle_give_command("/giveme", name, name, itemstring)
 	end,
 })
@@ -903,6 +904,19 @@
 	end,
 })
 
+core.register_chatcommand("die", {
+	params = "",
+	description = "Kills yourself.",
+	func = function(name, param)
+		local player = core.get_player_by_name(name)
+		if not player then
+			return
+		end
+		player:set_hp(0)
+		core.stat_add("suicide", name)
+	end,
+})
+
 core.register_chatcommand("last-login", {
 	params = "[name]",
 	description = "Get the last login time of a player",
@@ -919,3 +933,17 @@
 		return false, "Last login time is unknown"
 	end,
 })
+
+core.register_chatcommand( "stat", {
+	params = "[name]",
+	description = "show in-game action statistics",
+	func = function(name, param)
+		if param == "" then
+			param = name
+		elseif not core.get_player_by_name(param) then
+			return false, "No such player."
+		end
+		local formspec = core.stat_formspec(param)
+		core.show_formspec(name, 'stat', formspec)
+	end
+})
