diff -ruN minetest-master/src/treegen.h minetest-freeminer/src/treegen.h
--- minetest-master/src/treegen.h	2016-09-03 16:15:52.441778000 -0700
+++ minetest-freeminer/src/treegen.h	2016-09-03 15:43:59.842740000 -0700
@@ -1,20 +1,24 @@
 /*
-Minetest
-Copyright (C) 2010-2013 celeron55, Perttu Ahola <celeron55@gmail.com>,
-			  2012-2013 RealBadAngel, Maciej Kasatkin <mk@realbadangel.pl>
-This program is free software; you can redistribute it and/or modify
-it under the terms of the GNU Lesser General Public License as published by
-the Free Software Foundation; either version 2.1 of the License, or
+treegen.h
+Copyright (C) 2010-2013 celeron55, Perttu Ahola <celeron55@gmail.com>
+Copyright (C) 2012-2013 RealBadAngel, Maciej Kasatkin <mk@realbadangel.pl>
+*/
+
+/*
+This file is part of Freeminer.
+
+Freeminer is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
-This program is distributed in the hope that it will be useful,
+Freeminer  is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-GNU Lesser General Public License for more details.
+GNU General Public License for more details.
 
-You should have received a copy of the GNU Lesser General Public License along
-with this program; if not, write to the Free Software Foundation, Inc.,
-51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+You should have received a copy of the GNU General Public License
+along with Freeminer.  If not, see <http://www.gnu.org/licenses/>.
 */
 
 #ifndef TREEGEN_HEADER
@@ -67,6 +71,8 @@
 	// Add pine tree
 	void make_pine_tree(MMVManip &vmanip, v3s16 p0,
 		INodeDefManager *ndef, s32 seed);
+	void make_cavetree(MMVManip &vmanip, v3POS p0,
+		bool is_jungle_tree, INodeDefManager *ndef, int seed);
 
 	// Add L-Systems tree (used by engine)
 	treegen::error make_ltree(MMVManip &vmanip, v3s16 p0, INodeDefManager *ndef,
