diff -ruN minetest-master/builtin/game/init.lua minetest-freeminer/builtin/game/init.lua
--- minetest-master/builtin/game/init.lua	2016-09-03 16:15:52.661779000 -0700
+++ minetest-freeminer/builtin/game/init.lua	2016-09-03 15:43:59.794740000 -0700
@@ -17,7 +17,12 @@
 dofile(gamepath.."deprecated.lua")
 dofile(gamepath.."misc.lua")
 dofile(gamepath.."privileges.lua")
-dofile(gamepath.."auth.lua")
+if core.setting_getbool("auth_kv") then
+	dofile(gamepath.."fm_auth.lua")
+else
+	dofile(gamepath.."auth.lua")
+end
+dofile(gamepath.."stat.lua")
 dofile(gamepath.."chatcommands.lua")
 dofile(gamepath.."static_spawn.lua")
 dofile(gamepath.."detached_inventory.lua")
@@ -27,4 +32,8 @@
 dofile(gamepath.."forceloading.lua")
 dofile(gamepath.."statbars.lua")
 
+if core.setting_getbool("mod_debugging") then
+	dofile(gamepath.."mod_debugging.lua")
+end
+
 profiler = nil
