diff -ruN minetest-master/src/network/connection.h minetest-freeminer/src/network/connection.h
--- minetest-master/src/network/connection.h	2016-09-03 16:15:52.529778000 -0700
+++ minetest-freeminer/src/network/connection.h	2016-09-03 15:43:59.890740000 -0700
@@ -1,22 +1,33 @@
 /*
-Minetest
+connection.h
 Copyright (C) 2013 celeron55, Perttu Ahola <celeron55@gmail.com>
+*/
+
+/*
+This file is part of Freeminer.
 
-This program is free software; you can redistribute it and/or modify
-it under the terms of the GNU Lesser General Public License as published by
-the Free Software Foundation; either version 2.1 of the License, or
+Freeminer is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
-This program is distributed in the hope that it will be useful,
+Freeminer  is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-GNU Lesser General Public License for more details.
+GNU General Public License for more details.
 
-You should have received a copy of the GNU Lesser General Public License along
-with this program; if not, write to the Free Software Foundation, Inc.,
-51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+You should have received a copy of the GNU General Public License
+along with Freeminer.  If not, see <http://www.gnu.org/licenses/>.
 */
 
+#include "config.h"
+
+#if !MINETEST_PROTO
+#include "network/fm_connection.h"
+#else
+//Not used, keep for reduce MT merge conflicts
+
+
 #ifndef CONNECTION_HEADER
 #define CONNECTION_HEADER
 
@@ -34,6 +45,14 @@
 #include <list>
 #include <map>
 
+#include <fstream>
+
+extern std::ostream *dout_con_ptr;
+extern std::ostream *derr_con_ptr;
+#define dout_con (*dout_con_ptr)
+#define derr_con (*derr_con_ptr)
+
+
 class NetworkPacket;
 
 namespace con
@@ -345,7 +364,7 @@
 	RPBSearchResult findPacket(u16 seqnum);
 
 	std::list<BufferedPacket> m_list;
-	u32 m_list_size;
+	std::atomic_uint m_list_size;
 
 	u16 m_oldest_non_answered_ack;
 
@@ -603,12 +622,12 @@
 		This is called after the Peer has been inserted into the
 		Connection's peer container.
 	*/
-	virtual void peerAdded(Peer *peer) = 0;
+	virtual void peerAdded(u16 peer_id) = 0;
 	/*
 		This is called before the Peer has been removed from the
 		Connection's peer container.
 	*/
-	virtual void deletingPeer(Peer *peer, bool timeout) = 0;
+	virtual void deletingPeer(u16 peer_id, bool timeout) = 0;
 };
 
 class PeerHelper
@@ -1015,7 +1034,7 @@
 	void Connect(Address address);
 	bool Connected();
 	void Disconnect();
-	void Receive(NetworkPacket* pkt);
+	u32 Receive(NetworkPacket* pkt, int timeout = 0);
 	void Send(u16 peer_id, u8 channelnum, NetworkPacket* pkt, bool reliable);
 	u16 GetPeerID() { return m_peer_id; }
 	Address GetPeerAddress(u16 peer_id);
@@ -1059,7 +1078,7 @@
 
 	MutexedQueue<ConnectionEvent> m_event_queue;
 
-	u16 m_peer_id;
+	std::atomic_ushort m_peer_id;
 	u32 m_protocol_id;
 
 	std::map<u16, Peer*> m_peers;
@@ -1078,8 +1097,14 @@
 	bool m_shutting_down;
 
 	u16 m_next_remote_peer_id;
+
+//freeminer:
+public:
+	size_t events_size() { return m_event_queue.size(); }
 };
 
 } // namespace
 
 #endif
+
+#endif
