diff -ruN minetest-master/builtin/game/privileges.lua minetest-freeminer/builtin/game/privileges.lua
--- minetest-master/builtin/game/privileges.lua	2016-09-03 16:15:52.629779000 -0700
+++ minetest-freeminer/builtin/game/privileges.lua	2016-09-03 15:43:59.774740000 -0700
@@ -56,6 +56,6 @@
 core.register_privilege("rollback", "Can use the rollback functionality")
 core.register_privilege("zoom", {
 	description = "Can zoom the camera",
-	give_to_singleplayer = false,
+	give_to_singleplayer = true,
 })
 
