diff -ruN minetest-master/builtin/game/register.lua minetest-freeminer/builtin/game/register.lua
--- minetest-master/builtin/game/register.lua	2016-09-03 16:15:52.453778000 -0700
+++ minetest-freeminer/builtin/game/register.lua	2016-09-03 15:43:59.850740000 -0700
@@ -139,7 +139,7 @@
 	end
 
 	-- Flowing liquid uses param2
-	if itemdef.type == "node" and itemdef.liquidtype == "flowing" then
+	if itemdef.type == "node" and itemdef.liquidtype == "flowing" and itemdef.paramtype2 == nil then
 		itemdef.paramtype2 = "flowingliquid"
 	end
 
@@ -511,6 +511,39 @@
 core.registered_on_item_eats, core.register_on_item_eat = make_registration()
 core.registered_on_punchplayers, core.register_on_punchplayer = make_registration()
 
+minetest.register_on_joinplayer(function(player)
+	if minetest.is_singleplayer() then
+		return
+	end
+	local player_name =  player:get_player_name()
+	minetest.chat_send_all("*** " .. player_name .. " joined the game.")
+end)
+
+minetest.register_on_dieplayer(function(player)
+	local player_name =  player:get_player_name()
+	if minetest.is_singleplayer() then
+		player_name = "You"
+	end
+
+	-- Idea from https://github.com/4Evergreen4/death_messages
+	-- Death by lava
+	local nodename = minetest.get_node(player:getpos()).name
+	if nodename == "default:lava_source" or nodename == "default:lava_flowing" then
+		minetest.chat_send_all(player_name .. " melted into a ball of fire.")
+	-- Death by drowning
+	elseif nodename == "default:water_source" or nodename == "default:water_flowing" then
+		minetest.chat_send_all(player_name .. " ran out of air.")
+	--Death by fire
+	elseif nodename == "fire:basic_flame" then
+		minetest.chat_send_all(player_name .. " burned up.")
+	--Death by something else
+	else
+		--minetest.chat_send_all(player_name .. core.colorize("#bb0000", " died."))
+		minetest.chat_send_all(player_name .. " died.")
+	end
+
+end)
+
 --
 -- Compatibility for on_mapgen_init()
 --
