diff -ruN minetest-master/src/network/serveropcodes.cpp minetest-freeminer/src/network/serveropcodes.cpp
--- minetest-master/src/network/serveropcodes.cpp	2016-09-03 16:15:52.481778000 -0700
+++ minetest-freeminer/src/network/serveropcodes.cpp	2016-09-03 15:43:59.862740000 -0700
@@ -19,6 +19,7 @@
 */
 
 #include "serveropcodes.h"
+#include "../config.h"
 
 const static ToServerCommandHandler null_command_handler = { "TOSERVER_NULL", TOSERVER_STATE_ALL, &Server::handleCommand_Null };
 
@@ -92,7 +93,9 @@
 	{ "TOSERVER_RECEIVED_MEDIA",           TOSERVER_STATE_STARTUP, &Server::handleCommand_ReceivedMedia }, // 0x41
 	{ "TOSERVER_BREATH",                   TOSERVER_STATE_INGAME, &Server::handleCommand_Breath }, // 0x42
 	{ "TOSERVER_CLIENT_READY",             TOSERVER_STATE_STARTUP, &Server::handleCommand_ClientReady }, // 0x43
-	null_command_handler, // 0x44
+
+	{ "TOSERVER_DRAWCONTROL",              TOSERVER_STATE_STARTUP, &Server::handleCommand_Drawcontrol }, // 0x44
+
 	null_command_handler, // 0x45
 	null_command_handler, // 0x46
 	null_command_handler, // 0x47
