diff -ruN minetest-master/src/constants.h minetest-freeminer/src/constants.h
--- minetest-master/src/constants.h	2016-09-03 16:15:52.665779000 -0700
+++ minetest-freeminer/src/constants.h	2016-09-03 15:43:59.982741000 -0700
@@ -1,20 +1,23 @@
 /*
-Minetest
+constants.h
 Copyright (C) 2013 celeron55, Perttu Ahola <celeron55@gmail.com>
+*/
+
+/*
+This file is part of Freeminer.
 
-This program is free software; you can redistribute it and/or modify
-it under the terms of the GNU Lesser General Public License as published by
-the Free Software Foundation; either version 2.1 of the License, or
+Freeminer is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
-This program is distributed in the hope that it will be useful,
+Freeminer  is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-GNU Lesser General Public License for more details.
+GNU General Public License for more details.
 
-You should have received a copy of the GNU Lesser General Public License along
-with this program; if not, write to the Free Software Foundation, Inc.,
-51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+You should have received a copy of the GNU General Public License
+along with Freeminer.  If not, see <http://www.gnu.org/licenses/>.
 */
 
 #ifndef CONSTANTS_HEADER
@@ -46,7 +49,7 @@
 #define RESEND_TIMEOUT_MIN 0.1
 #define RESEND_TIMEOUT_MAX 3.0
 // resend_timeout = avg_rtt * this
-#define RESEND_TIMEOUT_FACTOR 4
+#define RESEND_TIMEOUT_FACTOR 8
 
 /*
     Server
@@ -73,9 +76,11 @@
 // results. (negative coordinates, values between nodes, ...)
 // Use floatToInt(p, BS) and intToFloat(p, BS).
 #define BS (10.0)
+#define HBS (BS/2)
 
 // Dimension of a MapBlock
-#define MAP_BLOCKSIZE 16
+#define MAP_BLOCKP 4
+#define MAP_BLOCKSIZE (1<<MAP_BLOCKP)
 // This makes mesh updates too slow, as many meshes are updated during
 // the main loop (related to TempMods and day/night)
 //#define MAP_BLOCKSIZE 32
@@ -90,14 +95,11 @@
 // Maximum hit points of a player
 #define PLAYER_MAX_HP 20
 
+#define PLAYER_FALL_TOLERANCE_SPEED (BS*14)
+
 // Maximal breath of a player
 #define PLAYER_MAX_BREATH 11
 
-// Number of different files to try to save a player to if the first fails
-// (because of a case-insensitive filesystem)
-// TODO: Use case-insensitive player names instead of this hack.
-#define PLAYER_FILE_ALTERNATE_TRIES 1000
-
 // For screenshots a serial number is appended to the filename + datetimestamp
 // if filename + datetimestamp is not unique.
 // This is the maximum number of attempts to try and add a serial to the end of
@@ -111,8 +113,10 @@
 // TODO: implement dpi-based scaling for windows and remove this hack
 #if defined(_WIN32)
 	#define TTF_DEFAULT_FONT_SIZE   (18)
+//#elif defined(__ANDROID__)
+//	#define TTF_DEFAULT_FONT_SIZE   (17)
 #else
-	#define TTF_DEFAULT_FONT_SIZE	(15)
+	#define TTF_DEFAULT_FONT_SIZE	(17)
 #endif
 #define DEFAULT_FONT_SIZE       (10)
 
